### ISSUE TITLE
Docker: Did not find a user after imputation finished

### ISSUE THREAD

**Issue by ad7718** [lassefolkersen/impute-me#43]
Imputation folder is filled with multiple files, but can't access data through the web interface.

**Comment by lassefolkersen** [comment]
Dear @ad7718 - docker running is still early stage, so we welcome any usage feedback very much. However, for this request it is necessary to ask for more details. The intended functionality is that you can access docker-generated data using one of these three approaches:
* by mounting a volume for the docker container, exposing the containers ~/data folder
* by downloading raw data through the docker-run web-interface using exposition of -p 3838, at http://localhost/www/<uniqueID> 
* by browsing calculated data through the docker-run web-interface, also with -p 3838, at http://localhost
which one is it that is causing trouble?

**Comment by ad7718** [comment]
I've uploaded data through web interface, then waited for imputation completion by monitoring running processes, and after that I tried to access (for example) Precision Medicine module through the the web and it can't find id. Also ~/data folder is empty, but imputation folder is full of imputed data and I can read it with text editor. I also tried to upload other data to process after that, but process is not running, I just see uploaded file and few others in imputation folder. Seems that I need to read some logs.

**Comment by lassefolkersen** [comment]
I see, thanks. In that case I suspect the run failed. Otherwise there definitely would be data in ~/data. That's something that probably should be improved with regards to error-messages in the future. But for now, could you please check the logs, with e.g. this command:

```
docker exec -it 8f39579bf867 find /home/ubuntu/logs/cron_logs -type f -exec cat {} +
```

Or just `find /home/ubuntu/logs/cron_logs -type f -exec cat {} +` if you are already inside the container. 

The current most common error we've seen, is that when docker-running on desktop-computers it needs to have the `max_imputation_chunk_size` in ~/misc_files/configuration.R-file entry tuned for using less RAM. Values down to around 500 seems to work better. But ok, I'm getting ahead of myself. Let's have a look at the logs first.

**Comment by ad7718** [comment]
File in attachement. Also I noticed that if I try to "find /home/ubuntu/logs/cron_logs -type f -exec cat {} +>/home/ubuntu/data/cron.log", there is no permission to do that, seems that's the problem?
[cron.log](https://github.com/lassefolkersen/impute-me/files/6375762/cron.log)

**Comment by lassefolkersen** [comment]
Ok, thanks. That did seem like it completed the entire imputation-run, so what I wrote before about RAM was not the issue. 
From the log-dump, the problem happens (first) at this line 
```
  cannot create dir '/home/ubuntu/data/id_82b4c5Q01', reason 'Permission denied'
 ```
The placement, means that it completed the imputation-procedure, but was prevented from moving the resulting files - and therefore could not complete the module-pre calculations (polygenic risk scores and such). 

I've not seen something like that before. 

Previously, I've done all docker-tests with this start command:
```
docker run -v ~/LOCALFOLDER:/home/ubuntu/data -p 3838:3838 lassefolkersen/impute-me:latest
```
Did you do something similar? Like, that ~/LOCALFOLDER - could that be blocked somehow?

**Comment by ad7718** [comment]
Yes, I've started docker as you described on dockerhub. Created ~/LOCALFOLDER by regular user and entered this command. This folder now is owned by root "drwxr-xr-x  2 root   root     4096 Apr 24 12:27 LOCALFOLDER" . I don't think I've created it by root, but I need to recheck.

**Comment by lassefolkersen** [comment]
Ok, sure - let me know if it works. Perhaps the instructions should also be changed so there is no need for docker to create a root-owned folder. Because the data-folder could equally well be somewhere guaranteed to exists, like ~. Or maybe write it so the user is sure to make their own folder, e.g. <FILL_IN_THIS> or something. I'm unsure what would be easiest/smartest, suggestions welcome.

**Comment by ad7718** [comment]
inside container:

ubuntu@6b381efa847e:~$ ls -la
total 52
drwxr-xr-x 1 ubuntu root   4096 Mar 29 18:29 .
drwxr-xr-x 1 root   root   4096 Mar 12 13:07 ..
-rw-r--r-- 1 ubuntu ubuntu   18 Mar 29 18:29 .bashrc
drwxr-xr-x 2 root   root   4096 Apr 26 14:11 data
drwxr-xr-x 2 ubuntu ubuntu 4096 Mar 12 13:07 imputations
drwxr-xr-x 5 ubuntu ubuntu 4096 Mar 12 13:07 logs
drwxr-xr-x 1 ubuntu ubuntu 4096 Mar 29 18:29 misc_files
drwxr-xr-x 1 ubuntu ubuntu 4096 Mar 12 13:15 programs
drwxr-xr-x 2 ubuntu ubuntu 4096 Mar 12 13:07 prs_dir
-rw-r--r-- 1 ubuntu ubuntu  108 Mar 29 18:29 .Rprofile
drwxr-xr-x 1 ubuntu ubuntu 4096 Mar 29 18:29 srv
drwxr-xr-x 2 ubuntu ubuntu 4096 Mar 12 13:07 vcfs
-rw-r--r-- 1 ubuntu ubuntu  172 Mar 12 13:15 .wget-hsts

**Comment by lassefolkersen** [comment]
Wait, I don't think it's the fault of Portainer... I've been playing around with this some more and I see the same behaviour. If the folder is not owned by user _ubuntu_ (outside of the container), it will become auto-owned by _root_ inside the container. Which it cannot. It must be owned by user _ubuntu_. So much for testing, if all my test-benches have users named only as _ubuntu_ lol.

I'll have to leave a full solution for later (or if some docker-expert reads this, please pipe in). But for now, I think the easiest is just to drop the `-v ~/LOCALFOLDER:/home/ubuntu/data` flag completely, when starting the docker. Because it's not really needed to expose a folder outside the container, anyway, not when that `-p` flag exposes the web-interface. All raw files can be gotten through that anyway (e.g. `http://localhost:3838/www/<uniqueID>_data.json` or `http://localhost:3838/www/<uniqueID>.gen.zip` in your browser, i.e. the option 2 and 3 that I listed above).

I'll update the instructions file right away. Thanks for all your feedback!

**Comment by lassefolkersen** [comment]
Edit: Actually - I'm still a bit confused by why this is, because it turned out we did test on some machines with non-_ubuntu_-users, e.g. a few Windows-machines. Anyway, the answer above should help your case @ad7718 - and more permanent solutions is pending better docker-insight (from someone)

**Comment by lassefolkersen** [comment]
Hi @ad7718 - just wanted to let you know that the latest release, v1.0.7, probably fixes many of the nuisances you've encountered. The nice people at [nf-core](https://nf-co.re) gave a lot of tips on portability. Maybe try again and see if the error is gone?
